algorithms/a3c/env_vizdoom.py

```diff
@@ -0,0 +1,379 @@
+
+        if self.params.use_train_instructions:
+            self.instructions = self.train_instructions
+        else:
+            self.instructions = self.test_instructions
+
+        self.word_to_idx = self.get_word_to_idx()
+        self.objects, self.object_dict = \
+            self.get_all_objects(os.path.join(self.curr_file_folder, self.params.all_instr_file))
+        self.object_sizes = self.read_size_file(os.path.join(self.curr_file_folder,
+                                                             self.params.object_size_file))
+        self.objects_info = self.get_objects_info()
+
+        # extra stuff to make it work with ai2thor A3C
+        self.task = Task()
+        self.observation_space = spaces.Box(low=0, high=255,
+                                            shape=(3, self.params.frame_height,
+                                                   self.params.frame_width),
+                                            dtype=np.uint8)
+        self.action_space = spaces.Discrete(3)
+        self.task.word_to_idx = self.word_to_idx
+
+    def game_init(self):
+        """Starts the doom game engine."""
+        game = DoomGame()
+        self.params.scenario_path = os.path.join(self.curr_file_folder, self.params.scenario_path)
+        game = set_doom_configuration(game, self.params)
+        game.init()
+        self.game = game
+
+    def reset(self):
+        """Starts a new episode.
+        Returns:
+           state: A tuple of screen buffer state and instruction.
+           reward: Reward at that step.
+        """
+
+        self.game.new_episode()
+        self.time = 0
+
+        self.instruction, instruction_id = self.get_random_instruction()
+
+        # Retrieve the possible correct objects for the instruction.
+        correct_objects = self.get_target_objects(instruction_id)
+
+        # Since we fix the number of objects to 5.
+        self.correct_location = np.random.randint(5)
+
+        # Randomly select one correct object from the
+        # list of possible correct objects for the instruction.
+        correct_object_id = np.random.choice(correct_objects)
+        chosen_correct_object = [x for x in self.objects_info if
+                                 x.name == self.objects[correct_object_id]][0]
+
+        # Special code to handle 'largest' and 'smallest' since we need to
+        # compute sizes for those particular instructions.
+        if 'largest' not in self.instruction \
+                and 'smallest' not in self.instruction:
+            object_ids = random.sample([x for x in range(len(self.objects))
+                                        if x not in correct_objects], 4)
+        else:
+            object_ids = self.get_candidate_objects_superlative_instr(
+                         chosen_correct_object)
+            object_ids = [self.object_dict[x] for x in object_ids]
+
+        assert len(object_ids) == 4
```

Comment:
Just a question to you, are you saying checks and raising an exception is always preferred to an assert?
The first could be 2-3 lines whereas an assert is often self-explanatory and 1 line

Reply:
Assert is for testing, I can ignore them whenever I want. To ignore an exception you need to change the code.
IMO the place for assertions is tests, but feel free to argue.

Reply:
Call Python with the -O flag to avoid asserts apparently. But why would someone do this with our file specifically?

asserts are concise. Raising Exceptions are ugly